### PR TITLE
Add environment setting to enable CUDA fast math on NVIDIA GPU.

### PR DESCRIPTION
**Pull request by JamesTheZ**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc [hunk 1]
```diff
@@ -305,9 +305,24 @@ class GpuKernelToBlobPass
...
+    // The default value is false currently because the UT of tf.FloorDiv UT
+    // fails if prec-div is not 2.
+    bool use_fast_math = false;
+    tensorflow::ReadBoolFromEnvVar("DISC_CUDA_USE_FAST_MATH", false,
+                                   &use_fast_math);
+    if (use_fast_math) {
+      // prec-div is set to 1 instead of 0 for better precision.
+      (*options.mutable_xla_backend_extra_options())["-nvptx-prec-divf32"] =
+          "1";
+      (*options.mutable_xla_backend_extra_options())["-nvptx-prec-sqrtf32"] =
```
Comment (minminsun) [inline tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc:317]: What about other math fuctions like exp, log, pow e.t.c.?

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc [hunk 2]
```diff
@@ -305,9 +305,24 @@ class GpuKernelToBlobPass
 
     xla::HloModuleConfig config;
     xla::DebugOptions options = xla::GetDebugOptionsFromFlags();
-    // Make sure we use full precision division operations.
-    // We need this to pass the ut for `tf.FloorDiv`.
-    (*options.mutable_xla_backend_extra_options())["-nvptx-prec-divf32"] = "2";
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc:310]: Is this is ok now? i remember there was a UT failure if we open this flag by default.

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc (hunk 1 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc:317]: They should be enabled by default. According to the document, the `use_fast_math` flag of `nvcc` only have these three additional settings.

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc (hunk 2 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc:310]: You remember right that this flag should be set to "2" for the UT of FloorDiv. It is moved to line-323, which is still the default setting.

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc [hunk 3]
```diff
@@ -305,9 +305,24 @@ class GpuKernelToBlobPass
...
+    // The default value is false currently because the UT of tf.FloorDiv UT
+    // fails if prec-div is not 2.
+    bool use_fast_math = false;
```
Comment (bddppq) [inline tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc:310]: I'm wondering if TF XLA enable fast math by default (we probably should follow the same default value as tf and pytorch)

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc [hunk 4]
```diff
@@ -305,9 +305,42 @@ class GpuKernelToBlobPass
...
+        // set to 1 instead of 0 for better precision.
+        (*options.mutable_xla_backend_extra_options())["-nvptx-prec-divf32"] =
+            "1";
+        (*options.mutable_xla_backend_extra_options())["-nvptx-prec-sqrtf32"] =
+            "0";
+        options.set_xla_gpu_ftz(true);
+        break;
+      case 3:
+        // Full fase-math for CUDA.
+        (*options.mutable_xla_backend_extra_options())["-nvptx-prec-divf32"] =
```
[the remark this answers is not in the document]
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc:335]: Fixed. Thanks.

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc [hunk 5]
```diff
@@ -328,6 +361,9 @@ class GpuKernelToBlobPass
             libdevice_dir, enable_fusion));
 
     VLOG(1) << "PTX code: \n" << ptx;
+#if 1
```
[the remark this answers is not in the document]
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc]: Removed.

File: tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc [hunk 6]
```diff
@@ -305,9 +305,42 @@ class GpuKernelToBlobPass
 
     xla::HloModuleConfig config;
     xla::DebugOptions options = xla::GetDebugOptionsFromFlags();
-    // Make sure we use full precision division operations.
-    // We need this to pass the ut for `tf.FloorDiv`.
-    (*options.mutable_xla_backend_extra_options())["-nvptx-prec-divf32"] = "2";
+    int64_t fast_math_level = 1;
+    tensorflow::ReadInt64FromEnvVar("DISC_CUDA_FAST_MATH_LEVEL", 1,
```
[the remark this answers is not in the document]
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_gpu_kernel_to_blob.cc:309]: I will make a list to explain all the existing env settings (e.g., stitch, codegen-schedule-opt, shape-constraint, fast-math) to users in a separate PR soon.

File: tao_compiler/mlir/disc/tests/tensorflow_ops/floor_div.cc [hunk 7]
```diff
@@ -24,6 +24,8 @@ const std::string c_ft_path =
 
 // static shape test case
 TEST(TFFloorDivOpTest, StaticShape2DF32) {
+  // Disable any fast-math setting for this UT.
+  setenv("DISC_CUDA_FAST_MATH_LEVEL", "0", 1);
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/tests/tensorflow_ops/floor_div.cc:28]: nit. You may need `SetUp`/`TearDown` from [here](https://github.com/google/googletest/blob/main/docs/primer.md#writing-the-main-function) instead set and unset env in each test case.

Comment (JamesTheZ) [reply tao_compiler/mlir/disc/tests/tensorflow_ops/floor_div.cc:28]: I will modify that in the next commit. Thanks.

**Pull request merged**